src/include/skiplist.h

```diff
@@ -0,0 +1,158 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+/*
+ * This file sets up our skip lists. In absence of C++ templates, we generate our own skip list
```

Comment:
If we have to do this for all the different algorithms, is it better just to switch to C++?

Reply:
That may be perhaps one motivation for C++, but it doesn't seem like we are anywhere close to making that switch. I agree that this syntax may seem somewhat unusual, but fortunately it's not that much more verbose than we would need for C++ templates. In any case, this approach (or using Python instead of the C preprocessor to turn the template into a running code) allows us to start getting the benefit of refactoring faster.

If we switch to C++, this would be an easy translation to the C++ template syntax, and it would result in a similar compilation result. C++ compilers implement templates by generating different functions and types (with different internal names) for each used combination of template parameters, which is basically what we are doing here, albeit more explicitly.

I also don't expect that we would need to do this for that many algorithms. Skip list is just particularly important, large (over ~600 lines of code if I move everything over), reusable with different types, and performance-critical. This would be an easier task for most other algorithms. For example, if an algorithm is less performance critical, we can refactor it using a different approach, such as how Linux kernel implements lists and RB trees.

Reply:
Thanks, make sense